sympy/geometry/line.py

```diff
@@ -663,141 +668,134 @@ def intersection(self, o):
+                v = Matrix([other.p1 - self.p1]).transpose()
+
+                # we cannot use m.solve(v) because that only works for square matrices
+                m_rref, pivots = m.col_insert(2, v).rref(simplify=True)
+                # rank == 2 ensures we have 2 pivots, but let's check anyway
+                if len(pivots) != 2:
+                    raise GeometryError("Failed when solving Mx=b when M={} and b={}".format(m,v))
+                coeff = m_rref[0,2]
+                line_intersection = self.direction*coeff + self.p1
+
+                # if we're both lines, we can skip a containment check
+                if isinstance(self, Line) and isinstance(other, Line):
+                    return [line_intersection]
+
+                if self.contains(line_intersection) and other.contains(line_intersection):
+                    return [line_intersection]
+                return []
+            else:
+                # we're skew
                 return []
 
-            # Not parallel, so find the point of intersection
-            px = simplify((b1*c2 - c1*b2) / t)
-            py = simplify((a2*c1 - a1*c2) / t)
-            inter = Point(px, py)
-            # we do not use a simplistic 'inter in self and inter in o'
-            # because that requires an equality test that is fragile;
-            # instead we employ some diagnostics to see if the intersection
-            # is valid
-
-            def inseg(self):
-                def _between(a, b, c):
-                    return c >= a and c <= b or c <= a and c >= b
-                if _between(self.p1.x, self.p2.x, inter.x) and \
-                        _between(self.p1.y, self.p2.y, inter.y):
-                    return True
-
-            def inray(self):
-                if self.p1 == inter:
-                    return True
-                sray = Ray(self.p1, inter)
-                if sray.xdirection == self.xdirection and \
-                        sray.ydirection == self.ydirection:
-                    return True
-
-            prec = (Line, Ray, Segment)
-            expr = self
-            if prec.index(expr.func) > prec.index(o.func):
-                expr, o = o, expr
-            rv = [inter]
-            if isinstance(expr, Line):
-                if isinstance(o, Line):
-                    return rv
-                elif isinstance(o, Ray) and inray(o):
-                    return rv
-                elif isinstance(o, Segment) and inseg(o):
-                    return rv
-            elif isinstance(expr, Ray) and inray(expr):
-                if isinstance(o, Ray) and inray(o):
-                    return rv
-                elif isinstance(o, Segment) and inseg(o):
-                    return rv
-            elif isinstance(expr, Segment) and inseg(expr):
-                if isinstance(o, Segment) and inseg(o):
-                    return rv
-            return []
-
-        return o.intersection(self)
+        return other.intersection(self)
```

Comment:
This is organized well and is clear. I leave the verification to the test suite.